testing/unittests/metrics/sinks/periodic/PeriodicSinkTests.cpp

```diff
@@ -0,0 +1,122 @@
+    See the License for the specific language governing permissions and
+    limitations under the License.
+############################################################################## */
+
+#ifdef _USE_CPPUNIT
+
+#include <cppunit/TestFixture.h>
+#include "metricunittests.hpp"
+#include <algorithm>
+
+#include "jptree.hpp"
+#include "jmetrics.hpp"
+
+#include "PeriodicTestSink.hpp"
+
+static MetricsReporter &reporter = queryMetricsReporter();
+
+const char *periodicSinkSettingsTestYml = R"!!(period: 2
+)!!";
+
+int period = 5;
+
+PeriodicTestSink *pPeriodicTestSink = nullptr;
+
+void usage()
+{
+    printf("\n"
+           "Usage:\n"
+           "    periodicsinktests>\n"
+           "\n");
+}
+
+int main(int argc, char* argv[])
+{
+    InitModuleObjects();
+    bool wasSuccessful = false;
+    {
+        // New scope as we need the TestRunner to be destroyed before unloading the dlls...
+        CppUnit::TestFactoryRegistry &registry = CppUnit::TestFactoryRegistry::getRegistry();
+        CppUnit::TextUi::TestRunner runner;
+        CppUnit::Test *all = registry.makeTest();
+        int numTests = all->getChildTestCount();
+        for (int i = 0; i < numTests; i++)
+        {
+            CppUnit::Test *sub = all->getChildTestAt(i);
+            runner.addTest(sub);
+        }
+        wasSuccessful = runner.run( "", false );
+    }
+    return wasSuccessful ? 0 : 1; // 0 == exit code success
+}
+
+class PeriodicSinkTests : public CppUnit::TestFixture
+{
+    public:
+        PeriodicSinkTests()
+        {
+            if (pPeriodicTestSink == nullptr)
+            {
+                //
+                // Load the settings then set the period using the global var
+                Owned<IPropertyTree> pSettings = createPTreeFromYAMLString(periodicSinkSettingsTestYml, ipt_none, ptr_ignoreWhiteSpace, nullptr);
+                pSettings->setPropInt("@period", period);
+                pPeriodicTestSink = new PeriodicTestSink("periodic_test_sink", pSettings);
+                reporter.addSink(pPeriodicTestSink, "periodic_test_sink");
+            }
+        }
+
+    CPPUNIT_TEST_SUITE(PeriodicSinkTests);
+        CPPUNIT_TEST(Test_sets_period_correctly);
+    CPPUNIT_TEST_SUITE_END();
+
+    protected:
+
+        void Test_sets_period_correctly()
+        {
+            //
+            // To test setting the period correctly, start collection and delay a multiple of that period.
+            // Stop collection and ask the test sink how many collections were done. If the count is +/- 1
+            // from the wait period multiple used, then we are close enough
+            int multiple = 5;
+            reporter.startCollecting();
+
+            //
+            // Check that the sink is collecting
+            CPPUNIT_ASSERT_MESSAGE("Expected sink to report it was collecting", pPeriodicTestSink->isCurrentlyCollectiing());
+
+            // wait... then stop collecing
+            sleep(multiple * period);
```

Comment:
This will is probably too long for a standard unit test.  May need to mark it as a stress test.